networkx/algorithms/d_separation.py

```diff
@@ -140,3 +140,202 @@ def d_separated(G, x, y, z):
+    of only the ancestors of X and Y. For full details, see [1]_.
+
+    Parameters
+    ----------
+    G : nx.DiGraph
+        The graph.
+    x : node
+        X node.
+    y : node
+        Y node.
+    z : Set
+        The separating set to check is minimal.
+
+    Returns
+    -------
+    bool
+        Whether or not the `z` separating set is minimal.
+
+    Raises
+    ------
+    NetworkXError
+        The *d-separation* test is commonly used with directed
+        graphical models which are acyclic.  Accordingly, the algorithm
+        raises a :exc:`NetworkXError` if the input graph is not a DAG.
+
+    NodeNotFound
+        If any of the input nodes are not found in the graph,
+        a :exc:`NodeNotFound` exception is raised.
+
+    References
+    ----------
+    .. [1] Tian, J., & Paz, A. (1998). Finding Minimal D-separators.
+
+    Notes
+    -----
+    This function only works on verifying a d-separating set is minimal
+    between two nodes. To verify that a d-separating set is minimal between
+    two sets of nodes is not supported.
+    """
+    if not nx.is_directed_acyclic_graph(G):
+        raise nx.NetworkXError("graph should be directed acyclic")
+
+    union_xy = set(x).union(set(y)).union(z)
+
+    if any(n not in G.nodes for n in union_xy):
+        raise nx.NodeNotFound("one or more specified nodes not found in the graph")
+
+    x_anc = nx.ancestors(G, x)
+    y_anc = nx.ancestors(G, y)
+    xy_anc = x_anc.union(y_anc)
+
+    # if Z contains any node which is not in ancestors of X or Y
+    # then it is definitely not minimal
+    if any(node not in xy_anc for node in z):
+        return False
+
+    D_anc_xy = x_anc.union(y_anc)
+    D_anc_xy = D_anc_xy.union((x, y))
+
+    # second, construct the moralization of the subgraph
+    moral_G = nx.moralize_graph(G.subgraph(D_anc_xy))
+
+    # start BFS from X
+    marks = _bfs_with_marks(moral_G, x, z)
+
+    # if not all the Z is marked, then the set is not minimal
+    if any(node not in marks for node in z):
+        return False
+
+    # similarly, start BFS from Y and check the marks
+    marks = _bfs_with_marks(moral_G, y, z)
+    # if not all the Z is marked, then the set is not minimal
+    if any(node not in marks for node in z):
+        return False
+
+    return True
+
+
+def _bfs_with_marks(G, start_node, check_set):
+    """Breadth-first-search with markings.
+
+    Performs BFS starting from ``start_node`` and whenever a node
+    inside ``check_set`` is met, it is "marked". Once a node is marked,
+    BFS does not continue along that path. The resulting marked nodes
+    are returned.
+
+    Parameters
+    ----------
+    G : nx.Graph
+        An undirected graph.
+    start_node : node
+        The start of the BFS.
+    check_set : set
+        The set of nodes to check against.
+
+    Returns
+    -------
+    marked : set
+        A set of nodes that were marked.
+    """
+    visited = dict()
+    marked = dict()
```

Comment:
I've used `dict` as a set-like object before too, but is there a reason here? It looks like the straight `set()` will do everything here and not require recasting at the end.  Changed involve this kind of syntax.
`visited.add(neighbr)`
`marked.add(neighbr)`
`return marked`

Reply:
Done.